src/sql/src/plan/statement/ddl.rs

```diff
@@ -1387,6 +1387,34 @@ pub fn plan_create_sink(
                         Ok(name_idx)
                     })
                     .collect::<Result<Vec<_>, _>>()?;
+                let is_valid_key =
+                    desc.typ().keys.iter().any(|key_columns| {
+                        key_columns.iter().all(|column| indices.contains(column))
+                    });
```

Comment:
Why is it necessary to do this check for non-upsert sources? I don't believe our other sources rely on the property of the Kafka key being a primary key.

Reply:
They don't rely on it, yes. I went for simplicity/uniformity but I think you're right that we shouldn't remove this allowed behaviour for other sinks.

Reply:
On second thought: does this make sense?

```
> CREATE MATERIALIZED SOURCE input_keyed2
  FROM KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'testdrive-input-keyed2-${testdrive.seed}'
  FORMAT AVRO USING SCHEMA '${schema}' ENVELOPE DEBEZIUM

> CREATE VIEW maxes2 AS SELECT a, max(b) AS b FROM input_keyed2 GROUP BY a

> CREATE SINK output_keyed2 FROM maxes2
  INTO KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'output-sink-keyed2-${testdrive.seed}'
  KEY (b)
  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY '${testdrive.schema-registry-url}' ENVELOPE DEBEZIUM

$ kafka-ingest format=avro topic=input-keyed2 schema=${schema} timestamp=1
{"before": null, "after": {"row": {"a": 1, "b": 1}}}
{"before": null, "after": {"row": {"a": 1, "b": 2}}}
{"before": null, "after": {"row": {"a": 1, "b": 3}}}

$ kafka-verify format=avro sink=materialize.public.output_keyed2
{"b": 3} {"before": null, "after": {"row": {"a": 1, "b": 3}}}

$ kafka-ingest format=avro topic=input-keyed2 schema=${schema} timestamp=1
{"before": {"row": {"a": 1, "b": 3}}, "after": null}
{"before": {"row": {"a": 1, "b": 2}}, "after": null}
{"before": null, "after": {"row": {"a": 2, "b": 1}}}

$ kafka-verify format=avro sink=materialize.public.output_keyed2
{"b": 1} {"before": null, "after": {"row": {"a": 1, "b": 1}}}
{"b": 1} {"before": null, "after": {"row": {"a": 2, "b": 1}}}
```

If you key by `a` the updates contained in the sink propose a different view of the world, which intuitively seems to make more sense to me:

```
> CREATE MATERIALIZED SOURCE input_keyed
  FROM KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'testdrive-input-keyed-${testdrive.seed}'
  FORMAT AVRO USING SCHEMA '${schema}' ENVELOPE DEBEZIUM

> CREATE VIEW maxes AS SELECT a, max(b) AS b FROM input_keyed GROUP BY a

> CREATE SINK output_keyed FROM maxes
  INTO KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'output-sink-keyed-${testdrive.seed}'
  KEY (a)
  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY '${testdrive.schema-registry-url}' ENVELOPE DEBEZIUM

$ kafka-ingest format=avro topic=input-keyed schema=${schema} timestamp=1
{"before": null, "after": {"row": {"a": 1, "b": 1}}}
{"before": null, "after": {"row": {"a": 1, "b": 2}}}
{"before": null, "after": {"row": {"a": 1, "b": 3}}}

$ kafka-verify format=avro sink=materialize.public.output_keyed
{"a": 1} {"before": null, "after": {"row": {"a": 1, "b": 3}}}

$ kafka-ingest format=avro topic=input-keyed schema=${schema} timestamp=1
{"before": {"row": {"a": 1, "b": 3}}, "after": null}
{"before": {"row": {"a": 1, "b": 2}}, "after": null}
{"before": null, "after": {"row": {"a": 2, "b": 1}}}

$ kafka-verify format=avro sink=materialize.public.output_keyed sort-messages=true
{"a": 2} {"before": null, "after": {"row": {"a": 2, "b": 1}}}
{"a": 1} {"before": {"row": {"a": 1, "b": 3}}, "after": {"row": {"a": 1, "b": 1}}}

```

Reply:
Ah sorry, the output for that first case is also correct: I forgot to check for the last update, which is:
```
$ kafka-verify format=avro sink=materialize.public.output_keyed2
{"b": 1} {"before": null, "after": {"row": {"a": 1, "b": 1}}}
{"b": 1} {"before": null, "after": {"row": {"a": 2, "b": 1}}}
{"b": 3} {"before": {"row": {"a": 1, "b": 3}}, "after": null}
```

It just feels a bit weird.

Reply:
Also: probably just ignore my last comments. I think it's good to just allow arbitrary keys for DEBEZIUM.